### ISSUE TITLE
Accessibility concerns for new fill-in-the-blank questions in English for Developers

### ISSUE THREAD

**Issue by bbsmooth** [freeCodeCamp/freeCodeCamp#52440]
### Describe the Issue

I'm creating this issue just to get the ball rolling to make sure that accessibility is taken into consideration as early as possible. I understand that these are still being worked on. This is just an initial assessment based on what is currently in the beta. I am using the "You Are" step in the "Learn Greetings in your First Day at the Office" for my assessment. 

- The first issue is that the text input in the sentence does not have an accessible name. Every input must have an accessible name. I think in this case it would make sense to assign the name with `aria-label`. As to what that name should be, I'll revisit in a bit.

- The second issue is that I don't think the instructions are clear enough. Yes, there is a heading that says "Fill in the blank", but I think there might need to be more. I think it should be made clear what value is expected in the blank. In this example step, we could add instructions below the heading that clarify exactly what value is expected. Perhaps something like:

> Add the appropriate form of the verb `to be` to the following sentence in order to introduce yourself to the new graphic designer.

Adding this additional information now potentially gives us something to use for the accessible name on the input. Since we have told the user that they should add a verb, we could use "verb" as the accessible name. Another option would be to use the word "blank" for the accessible name and reword the instructions a little:

> Add the appropriate form of the verb `to be` in the blank input in the following sentence in order to introduce yourself to the new graphic designer.

As to which method is better, I think this is something that we really need to get feedback from native screen reader users. Have we ever put a call out to screen reader users before to help with testing? I'd be willing to put together some examples for testing.

- Third issue: Not sure if this was meant to stay, but we should get rid of the "Click the button below to check your answer" sentence. I think having a button that says "Check your answer" is clear enough. This sentence is basically just redundant noise.

- Fourth issue: When the user checks their answer and it is wrong, the error message needs to be automatically announced to screen reader users. If you look at the `div` we put the error message in for RWD you'll see that is has an `aria-live` set to "polite". This causes whatever text is placed in that `div` to be automatically announced by a screen reader. Another option would be to put the error message in a `p` element with `tabindex="-1"` and then programmatically set the focus on it after the text has been added. I think this might be the better approach if we are going to leave the error message above the check answer button. I also think the error message needs to be restructured a little. The first thing that should be announced is that the answer is not correct. Then give an explanation of why it is not correct. And then finally ask them to try again.

- Fifth issue: I know there are discussions about creating a custom audio player for these steps. I think it would be a good idea to make sure it has an accessible name, or perhaps put it in a region landmark with an accessible name so it is easier for screen reader users to jump to. We might also want to think about adding a reference to it in the instructions. For example:

> Listen to the following audio clip to hear how Maria uses the verb `to be` to introduce herself to Tom and confirm his job role.

And then of course, all audio clips will need a transcript. I think the easiest way to provide this is with a disclosure widget immediately below the audio player. It could be a "transcript" button that toggles the display of the transcript below it. Or we could use `details`/`summary` elements.

### Affected Page

http://localhost:8000/learn/a2-english-for-developers/learn-greetings-in-your-first-day-at-the-office/you-are

### Your code

N/A

### Expected behavior

N/A

### System

N/A

**Comment by bbsmooth** [comment]
> we are likely going to change all of the English challenges to remove the audio element, and replace it with an animation (picture with audio)

Sounds good. If the animations themselves aren't going to add any valuable information and are rather just acting as eye-candy, and the real information is coming from just the audio, then I don't foresee any additional accessibility issues here. We most likely won't need audio descriptions because the video isn't providing any information that needs to be described. So I'll just keep referring to these as "audio" even though there is going to be background video.

I understand what you are saying about transcripts for the audio giving away the answer. I think we might be able to work around this depending on how the audio is going to be used. Correct me if I'm wrong, but it seems like the audio is being used just to speak the fill in the blank sentence (but with the blank filled in). The user can then listen to the full sentence and fill in the blank based on what they heard. But it wouldn't be mandatory to listen to the audio if we gave them clear instructions instead, such as I suggested above. In other words, the instructions would act as a reasonable alternative to the audio and then we wouldn't run into any WCAG issues. This would also mandate that the content before the audio completely explains any concepts in the audio. In other words, **if we write the content so that the audio can be removed and no information would be lost**, then I think we are doing pretty good as far as accessibility is concerned.

> Other sites have literally no instructions.

If there is one thing to know about accessibility professionals, we put very little faith in what other sites are doing because even the ones who should know better often don't get it right (either accidentally or intentionally) :slightly_smiling_face: But if you can point me to some of these sights I would much appreciate it.

As far as trying to find volunteer screen reader users, I was just wondering what would be the best way to do this? Could we create a post in the forum? Quincy's got a huge mailing list, maybe he would be willing to ask for volunteers in one of his weekly emails? Could we add something to the `/learn` landing page? I'm just throwing out ideas here. I have no idea what's feasible. We could also just try our best and wait for feedback. I just prefer to be a little more pro-active.

**Comment by bbsmooth** [comment]
I finally had a chance to check out clozemaster. It does have accessibility issues, but the one big difference between what it is doing and what we are doing here is that it is basically giving you complete instructions since it gives you the entire sentence you need to translate and then you just have to fill in the missing word in the translation. So in essence, it is giving you instructions on what to do (i.e. it is giving you the sentence to translate). On the other hand, we aren't giving the user the complete sentence, we are giving them everything except a word and asking them to fill it in. Thus, someone who can't hear the video won't get the complete instructions on how to complete the fill in the blank. So we need to make that clear with alternative instructions.

**Comment by moT01** [comment]
✅ every input must have an accessible name

> add instructions below the heading that clarify exactly what value is expected

I think this is going to be tough. For challenges with one blank, it would work - but some challenges now have up to 10 blanks or so. So I don't think it makes sense to add this.

✅ get rid of the "Click the button below to check your answer"

[ ] announce feedback - [github.com/freeCodeCamp/freeCodeCamp/pull/56822](https://github.com/freeCodeCamp/freeCodeCamp/pull/56822) 

[ ] triage the rest of this issue

**Comment by bbsmooth** [comment]
"some challenges now have up to 10 blanks or so"

@moT01 Can you tell me where I can find one of these?

**Comment by ojeytonwilliams** [comment]
Hey Bruce 👋. Here's one I found with nine blanks: https://www.freecodecamp.org/learn/a2-english-for-developers/learn-how-to-read-and-understand-code-documentation/task-15

**Comment by moT01** [comment]
Maybe some generic instructions? "Fill in the blanks in the below with the words from the dialogue."

**Comment by jdwilkin4** [comment]
@moT01 

What's the status for this one?

**Comment by jdwilkin4** [comment]
I am going to remove the `discussing` label since there have been no active discussions for a while.

**Comment by moT01** [comment]
I think we're safe to close this given the lack of activity. We've fixed many of the issues and things are quite accessible. If anyone still has any concerns, It might be best to create a new issue to try and start up a new conversation.
